File: components/astraRGBD/src/specificworker.cpp
/*
 *    Copyright (C)2018 by YOUR NAME HERE
 *
 *    This file is part of RoboComp
 *
 *    RoboComp is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    RoboComp is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "specificworker.h"

/**
* \brief Default constructor
*/
SpecificWorker::SpecificWorker(MapPrx& mprx) : GenericWorker(mprx)
{
}

/**
* \brief Default destructor
*/
SpecificWorker::~SpecificWorker()
{
       astra::terminate();
}

bool SpecificWorker::setParams(RoboCompCommonBehavior::ParameterList params)
{
//       THE FOLLOWING IS JUST AN EXAMPLE
//
//	try
//	{
//		RoboCompCommonBehavior::Parameter par = params.at("InnerModelPath");
//		innermodel_path = par.value;
//		innermodel = new InnerModel(innermodel_path);
//	}
//	catch(std::exception e) { qFatal("Error reading config params"); }

    depthB = QString::fromStdString(params["depth"].value).contains("true");
	colorB = QString::fromStdString(params["color"].value).contains("true");



	timer.start(Period);
    astra::initialize();
    initializeStreams();



	return true;
}

void SpecificWorker::compute()
{
	QMutexLocker locker(mutex);
	//computeCODE
// 	try
// 	{
// 		camera_proxy->getYImage(0,img, cState, bState);
// 		memcpy(image_gray.data, &img[0], m_width*m_height*sizeof(uchar));
// 		searchTags(image_gray);
// 	}
// 	catch(const Ice::Exception &e)
// 	{
// 		std::cout << "Error reading from Camera" << e << std::endl;
// 	}

    readFrame();

}


Registration SpecificWorker::getRegistration()
{
//implementCODE

}

void SpecificWorker::getData(imgType &rgbMatrix, depthType &distanceMatrix, RoboCompJointMotor::MotorStateMap &hState, RoboCompGenericBase::TBaseState &bState)
{
//implementCODE
	RGBMutex->lock();
	rgbMatrix=*colorImage;
	RGBMutex->unlock();
	depthMutex->lock();
	distanceMatrix=*depthImage;
	depthMutex->unlock();

}

void SpecificWorker::getXYZ(PointSeq &points, RoboCompJointMotor::MotorStateMap &hState, RoboCompGenericBase::TBaseState &bState)
{
//implementCODE

}

void SpecificWorker::getRGB(ColorSeq &color, RoboCompJointMotor::MotorStateMap &hState, RoboCompGenericBase::TBaseState &bState)
{
//implementCODE

}

TRGBDParams SpecificWorker::getRGBDParams()
{
//implementCODE

}

void SpecificWorker::getDepth(DepthSeq &depth, RoboCompJointMotor::MotorStateMap &hState, RoboCompGenericBase::TBaseState &bState)
{
//implementCODE

}

void SpecificWorker::setRegistration(const Registration &value)
{
//implementCODE

}

void SpecificWorker::getImage(ColorSeq &color, DepthSeq &depth, PointSeq &points, RoboCompJointMotor::MotorStateMap &hState, RoboCompGenericBase::TBaseState &bState)
{
//implementCODE

}

void SpecificWorker::getDepthInIR(depthType &distanceMatrix, RoboCompJointMotor::MotorStateMap &hState, RoboCompGenericBase::TBaseState &bState)
{
//implementCODE

}


void SpecificWorker::initializeStreams()
{
    qDebug()<<"\tInitializing streams"<<endl;
	if (depthB) {
        depthReader = new(astra::StreamReader);
        qDebug()<<"\t\tDepth reader created"<<endl;
        *depthReader = streamSet.create_reader();

        depthReader->stream<astra::DepthStream>().start();
        astra::Frame frame = depthReader->get_latest_frame();
        const auto depthFrame = frame.get<astra::DepthFrame>();
        int width = depthFrame.width();
        int height = depthFrame.height();
        depthImage = new vector<float>(width*height);
        qDebug()<<"\t\tDepth stream initalized";
	}

	if (colorB){
	    qDebug()<<"\t\tColor reader created"<<endl;
	    *colorReader = streamSet.create_reader();
        colorReader->stream<astra::ColorStream>().start();
        astra::Frame frame = depthReader->get_latest_frame();
        const auto colorFrame = frame.get<astra::ColorFrame>();
	    int width = colorFrame.width();
        int height = colorFrame.height();
        colorImage = new vector<Ice::Byte>(width*height*3);
	}
}

void SpecificWorker::readFrame()
{
    astra::Frame frame = depthReader->get_latest_frame();
    fps++;
	if (depthB) {

        const auto depthFrame = frame.get<astra::DepthFrame>();
        int width = depthFrame.width();
        int height = depthFrame.height();
        const int16_t* depthPtr = depthFrame.data();
        RGBMutex->lock();
    	memcpy(&depthImage->operator[](0),depthPtr,width*height);
	    RGBMutex->unlock();
	}

	if (colorB){
	    const auto colorFrame = frame.get<astra::ColorFrame>();
	    int width = colorFrame.width();
        int height = colorFrame.height();
        const astra::RgbPixel* colorPtr = colorFrame.data();
        RGBMutex->lock();
    	memcpy(&colorImage->operator[](0),colorPtr,width*height*3);
	    RGBMutex->unlock();
	}

}

File: components/detection/test/imageBasedGestureRecognitionClient/src/specificworker.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#
#    Copyright (C) 2021 by YOUR NAME HERE
#
#    This file is part of RoboComp
#
#    RoboComp is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    RoboComp is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
#

from rich.console import Console
from genericworker import *
from RoboCompImageBasedGestureRecognition import *
import traceback
import numpy as np
import time
import cv2

sys.path.append('/opt/robocomp/lib')
console = Console(highlight=False)


# If RoboComp was compiled with Python bindings you can use InnerModel in Python
# import librobocomp_qmat
# import librobocomp_osgviewer
# import librobocomp_innermodel

class Timer:
    def __init__(self, fps):
        self.fps = fps
        self.ms = 1000 / fps
        self.tick = None

    # tock - current time in milliseconds
    def isReady(self, tock):
        if self.tick is None or tock - self.tick > self.ms:
            self.tick = tock
            return True
        return False


current_milli_time = lambda: int(round(time.time() * 1000))


class SpecificWorker(GenericWorker):
    def __init__(self, proxy_map, startup_check=False):
        super(SpecificWorker, self).__init__(proxy_map)
        self.timer.timeout.connect(self.compute)
        self.Period = 30
        self.timer.start(self.Period)
        # self.visualizer = Visualizer()
        self.cam_timer = Timer(fps=30)
        self.inference_timer = Timer(fps=10)
        self.capL = cv2.VideoCapture(0)

        self.max_num_images = 64
        self.list_frames = []

    def __del__(self):
        console.print('SpecificWorker destructor')

    def setParams(self, params):
        return True

    @QtCore.Slot()
    def compute(self):
        now = current_milli_time()
        cam_ready = self.cam_timer.isReady(now)
        if cam_ready:
            retL, self.frameL = self.capL.read()
            self.list_frames.append(self.frameL)

        if self.inference_timer.isReady(now) and len(self.list_frames) >= self.max_num_images:
            input = TVideo()
            input.images = np.stack(self.list_frames, axis=0)
            input.height, input.width, input.depth = self.list_frames[0].shape
            input.numFrames = len(self.list_frames)
            self.list_frames = []
            output = self.imagebasedgesturerecognition_proxy.getGesture(input)

            print(output.gestureIndex)
            print(output.gestureProb)

        return True

    def startup_check(self):
        import time
        time.sleep(0.2)
        exit()

    ######################
    # From the RoboCompCameraSimple you can call this methods:
    # self.camerasimple_proxy.getImage(...)

    ######################
    # From the RoboCompCameraSimple you can use this types:
    # RoboCompCameraSimple.TImage

    ######################
    # From the RoboCompImageBasedGestureRecognition you can call this methods:
    # self.imagebasedgesturerecognition_proxy.getGesture(...)

    ######################
    # From the RoboCompImageBasedGestureRecognition you can use this types:
    # RoboCompImageBasedGestureRecognition.TVideo
    # RoboCompImageBasedGestureRecognition.GestureResult
